### ISSUE TITLE
Invoking glob() inside completion function breaks <C-P> functionality

### ISSUE THREAD

**Issue by svilenb** [neovim/neovim#10563]
- `nvim --version`: v0.4.0-dev
- Operating system/version:  Ubuntu 18.04.2 LTS

I'm opening this issue as a result of the [following](https://github.com/tpope/vim-projectionist/issues/132) discussion.

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
:e test_complete.vim
```
Put the following content into that test_complete.vim file and write it:

```
function! TestComplete(A, L, P) abort
  return ['a1', 'a2']
endfunction

command! -nargs=* -complete=customlist,TestComplete TestComplete echo <q-args>

function! TestGlobComplete(A, L, P) abort
  call glob('a*')
  return ['a1', 'a2']
endfunction

command! -nargs=* -complete=customlist,TestGlobComplete TestGlobComplete echo <q-args>
```
Source the newly created file and test the two functions:

```
:so test_complete.vim
:TestComplete a<Tab>
:TestGlobComplete a<Tab>
```
### Actual behaviour

For the first `:TestComplete a<Tab> `command you can use `CTRL-P` and `CTRL-N` to cycle through the matches, eventually ending up back to what was originally type (in this case it's just "a").

For the second `:TestGlobComplete a<Tab>` command you cannot go straight back to what originally was typed with `CTRL-P` and `CTRL-N`. This means that you are always forced to choose one of the completion results.

### Expected behaviour

Calling glob() inside the second TestGlobComplete completion function shouldn't remove the ability to move away from the completion results.

**Comment by justinmk** [comment]
cc @chrisbra @brammool

> invoking the glob() function anywhere inside completion function removes the ability to CTRL-P away the completion result, even if we completely ignore the return value.

**Comment by svilenb** [comment]
The newly added <C-E> behavior from [#11746](https://github.com/neovim/neovim/pull/11746) is also broken in the case of `:TestGlobComplete a<Tab>`. The following error is show when using <C-E>:

>  9722 segmentation fault (core dumped)   nvim

**Comment by zeertzjq** [comment]
This seems to be caused by the `orig_save` static variable:
```c
char *ExpandOne(expand_T *xp, char *str, char *orig, int options, int mode)
{
  char *ss = NULL;
  static char *orig_save = NULL;      // kept value of orig
  int orig_saved = false;
```

**Comment by zeertzjq** [comment]
Fixed by #10563
